### ISSUE TITLE
Running tests in parallel sporadically fails with exit code 255

### ISSUE THREAD

**Issue by asendimitrov** [nunit/nunit-console#705]
Hi, I'm using NUnit 24/7 at work and I'm sporadically hitting an issue when running tests in parallel.

NUnit Framework version: 3.12 (NuGet)
NUnit ConsoleRunner version: 3.10 (NuGet)
Console command: `%nunit.console.path% --trace=Debug --inprocess --result=%nunit.result.path% %tests.dlls.list% --workers=10`
ParallelizableScope: Children (on assembly level)

The issue occurs around 10% of the time on our TeamCity agents and the error is `Process exited with code 255`. It _looks to be affected_ by the hardware of the box the tests are ran on - there's no hard proof of this - but I've never managed to reproduce it locally (the agents are VMs with weakish dual core processors whereas mine has 8).

Lowering the `workers`  doesn't seem to affect as I've seen the same result with both `5` and not specified at all (would default to 2 when ran on TC).

I'm always using `--inprocess` ever since I first hit the issue in order to see the real reason why this happens but there's nothing specific either as an exception by the console or in the logs. Omitting `--inprocess` results in a SocketException issue (#255 )

What I've noticed is that this (mostly) occurrs when test number X is running, where X is equal to the number of workers. Here's an example with both 5 and 10:

5 workers:
```
[11:38:46]MyCompany.dll (23s)
[11:38:53]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("1") (16s)
[11:38:53]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("2") (16s)
[11:38:53]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("3") (16s)
[11:38:53]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("4") (16s)
[11:38:53]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("5") (16s)
[11:39:10]Process exited with code 255
[11:39:10]Process exited with code 255 (Step: Run tests (Command Line))
[11:39:10]Step Run tests (Command Line) failed
```
logs: [5 workers - InternalTrace.5864.zip](https://github.com/nunit/nunit-console/files/3808969/5.workers.-.InternalTrace.5864.zip)

10 workers:
```
[11:03:42]MyCompany.dll (11s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("1") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("2") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("3") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("4") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("5") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test1("6") (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test2(50,111.07d,50.1d) (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test3("1", 50,41.5d) (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test3("2", 50,41.5d) (5s)
[11:03:48]MyCompany.Tests.GetBalances.GetBalancesTests.Test4(50,"Default") (5s)
[11:03:53]Process exited with code 255
[11:03:53]Process exited with code 255 (Step: Run tests (Command Line))
[11:03:53]Step Run tests (Command Line) failed
```
logs:
[10 workers - InternalTrace.2948.zip](https://github.com/nunit/nunit-console/files/3808994/10.workers.-.InternalTrace.2948.zip)

Please let me know if you need anything else. Thanks.

**Comment by ChrisMaddock** [comment]
There's nothing in the logs to help here, unfortunately. Do you get anything additional logged when you run without `--inprocess` - other than the SocketException, of course?

**Comment by asendimitrov** [comment]
I don't think anything unusual other than default SocketException appears in the logs in that case.
[SocketException - InternalTrace.2136.zip](https://github.com/nunit/nunit-console/files/3814366/SocketException.-.InternalTrace.2136.zip)

**Comment by ChrisMaddock** [comment]
You're right, there's no extra clues here. I'm afraid I'm really not sure what would cause this, perhaps others on @nunit/engine-team might have some ideas?

**Comment by jnm2** [comment]
The SocketException means the agent process died unceremoniously. What's the error message when the agent is not in the picture (use `--inprocess`)?

**Comment by ChrisMaddock** [comment]
Nothing, it seems - see the top post. That's the bit that's confusing me! 🙂

**Comment by jnm2** [comment]
Oh, sorry. This is a long shot but I wonder if it could possibly be fixed by https://github.com/nunit/nunit-console/pull/660. How about trying the latest from the prerelease feed?

**Comment by jnm2** [comment]
The internal trace just stops, similar to what I was seeing with the .NET Framework race condition that #660 avoids.

(Btw, the race condition is fixed in the .NET Framework November updates.)

**Comment by asendimitrov** [comment]
Doesn't seem to help in any way :(

Tried several times with `NUnit.ConsoleRunner.3.11.0-dev-04648` and results are exactly the same: logs contain just the SocketException or nothing at all if running with `--inprocess`

**Comment by CharliePoole** [comment]
@nunit/engine-team Doesn't the engine get distributed with access to source code for debugging? Maybe we should make `--agent-debug` available in release builds so people can step through it?

@asendimitrov Just to be sure... have you looked at the separate internal trace log produced for the agent process? Each run may give you three or four logs.

**Comment by asendimitrov** [comment]
@CharliePoole thanks for tuning in. Yes, I've checked all 3 log files and there doesn't see to be a hint in either of them. Above I've shared a zip containing the nunit agent process log, and in it the last messages are: 
```
12:08:40.441 Info  [ 5] NUnitFrameworkDriver: Loading MyCompany.dll - see separate log file
12:08:40.661 Info  [ 5] NUnitFrameworkDriver: Loaded MyCompany.dll
12:08:40.662 Info  [ 5] NUnitFrameworkDriver: Running MyCompany.dll - see separate log file
```
The process breaks after those timestamps, unless I'm missing anything?
